File: app/src/main/java/com/signbridgecommunication/app/viewmodel/MainViewModel.java
package com.signbridgecommunication.app.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.signbridgecommunication.app.data.datastore.OnboardingManager;
import com.signbridgecommunication.app.data.model.Category;
import com.signbridgecommunication.app.data.model.Phrase;
import com.signbridgecommunication.app.data.repository.SignBridgeRepository;
import java.util.List;
import java.util.Map;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;

public class MainViewModel extends ViewModel {
    private final SignBridgeRepository repository;
    private final OnboardingManager onboardingManager;
    private final MutableLiveData<Boolean> isOnboardingCompleted = new MutableLiveData<>();
    private final CompositeDisposable disposables = new CompositeDisposable();

    public MainViewModel(SignBridgeRepository repository, OnboardingManager onboardingManager) {
        this.repository = repository;
        this.onboardingManager = onboardingManager;
        
        disposables.add(onboardingManager.isOnboardingCompleted()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(isOnboardingCompleted::setValue));
    }

    public LiveData<Boolean> isOnboardingCompleted() {
        return isOnboardingCompleted;
    }

    public LiveData<List<Category>> getCategories() {
        return repository.getCategories();
    }

    public LiveData<List<Phrase>> getFavoritePhrases() {
        return repository.getFavoritePhrases();
    }

    public LiveData<List<Phrase>> getRecentPhrases() {
        return repository.getRecentPhrases();
    }

    public LiveData<Map<String, Integer>> getStats() {
        return repository.getStats();
    }

    public void completeOnboarding() {
        onboardingManager.setOnboardingCompleted(true);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposables.clear();
    }
}

File: app/src/main/java/com/signbridgecommunication/app/viewmodel/MainViewModel.kt
package com.signbridgecommunication.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.signbridgecommunication.app.data.datastore.OnboardingDataStore
import com.signbridgecommunication.app.data.model.Category
import com.signbridgecommunication.app.data.model.Phrase
import com.signbridgecommunication.app.data.repository.SignBridgeRepository
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class MainViewModel(
    private val repository: SignBridgeRepository,
    private val onboardingDataStore: OnboardingDataStore
) : ViewModel() {

    val isOnboardingCompleted = onboardingDataStore.isOnboardingCompleted
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), false)

    val categories = repository.getCategories()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val favoritePhrases = repository.getFavoritePhrases()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val recentPhrases = repository.getRecentPhrases()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val stats = repository.getStats()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyMap())

    private val _searchQuery = MutableStateFlow("")
    val searchQuery = _searchQuery.asStateFlow()

    @OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
    val searchResults = _searchQuery
        .debounce(300)
        .flatMapLatest { query ->
            if (query.isBlank()) flowOf(emptyList())
            else repository.searchPhrases(query)
        }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
    }

    fun completeOnboarding() {
        viewModelScope.launch {
            onboardingDataStore.saveOnboardingCompleted(true)
        }
    }
}
